File: src/main/java/com/github/samarium150/structurescompass/gui/TransparentTextField.java
package com.github.samarium150.structurescompass.gui;

import com.github.samarium150.structurescompass.util.GeneralUtils;
import com.github.samarium150.structurescompass.util.RenderUtils;
import com.mojang.blaze3d.matrix.MatrixStack;
import com.mojang.blaze3d.platform.GlStateManager;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.widget.TextFieldWidget;
import net.minecraft.client.renderer.BufferBuilder;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

import javax.annotation.Nonnull;

/**
 * Class of the transparent style text field
 * <p>
 * This class is adapted from
 * <a href="https://github.com/MattCzyr/NaturesCompass" target="_blank">NaturesCompass</a>
 * which is under
 * <a href="https://creativecommons.org/licenses/by-nc-sa/4.0" target="_blank">
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
 * </a>.
 */
@OnlyIn(Dist.CLIENT)
public class TransparentTextField extends TextFieldWidget {
    
    private final FontRenderer fontRenderer;
    private ITextComponent label;
    private int labelColor = 0x808080;
    
    private boolean pseudoIsEnabled = true;
    private boolean pseudoEnableBackgroundDrawing = true;
    private int pseudoMaxStringLength = 32;
    private int pseudoLineScrollOffset;
    private int pseudoEnabledColor = 14737632;
    private int pseudoDisabledColor = 7368816;
    private int pseudoCursorCounter;
    private int pseudoSelectionEnd;
    
    public TransparentTextField(FontRenderer fontRenderer, int x, int y, int width, int height, ITextComponent label) {
        super(fontRenderer, x, y, width, height, label);
        this.fontRenderer = fontRenderer;
        this.label = label;
    }
    
    @Override
    public void render(@Nonnull MatrixStack matrixStack, int mouseX, int mouseY, float partialTicks) {
        if (isVisible()) {
            if (pseudoEnableBackgroundDrawing) {
                final int color = (int) (255.0F * 1.7F);
                RenderUtils.drawRect(x, y, x + width, y + height, color / 2 << 24);
            }
            boolean showLabel = !isFocused() && getValue().isEmpty();
            int i = showLabel ? labelColor : (pseudoIsEnabled ? pseudoEnabledColor : pseudoDisabledColor);
            int j = getCursorPosition() - pseudoLineScrollOffset;
            int k = pseudoSelectionEnd - pseudoLineScrollOffset;
            String text = showLabel ? label.getString() : getValue();
            String s = fontRenderer.plainSubstrByWidth(text.substring(pseudoLineScrollOffset), getWidth());
            boolean flag = j >= 0 && j <= s.length();
            boolean flag1 = isFocused() && pseudoCursorCounter / 6 % 2 == 0 && flag;
            int l = pseudoEnableBackgroundDrawing ? x + 4 : x;
            int i1 = pseudoEnableBackgroundDrawing ? y + (height - 8) / 2 : y;
            int j1 = l;
            
            if (k > s.length()) {
                k = s.length();
            }
            
            if (!s.isEmpty()) {
                String s1 = flag ? s.substring(0, j) : s;
                j1 = fontRenderer.drawShadow(matrixStack, s1, (float) l, (float) i1, i);
            }
            
            boolean flag2 = getCursorPosition() < getValue().length() || getValue().length() >= pseudoMaxStringLength;
            int k1 = j1;
            
            if (!flag) {
                k1 = j > 0 ? l + width : l;
            } else if (flag2) {
                k1 = j1 - 1;
                --j1;
            }
            
            if (!s.isEmpty() && flag && j < s.length()) {
                fontRenderer.drawShadow(matrixStack, s.substring(j), (float) j1, (float) i1, i);
            }
            
            if (flag1) {
                if (flag2) {
                    RenderUtils.drawRect(k1, i1 - 1, k1 + 1, i1 + 1 + fontRenderer.lineHeight, -3092272);
                } else {
                    fontRenderer.drawShadow(matrixStack, "_", (float) k1, (float) i1, i);
                }
            }
            
            if (k != j) {
                int l1 = l + fontRenderer.width(s.substring(0, k));
                drawSelectionBox(k1, i1 - 1, l1 - 1, i1 + 1 + fontRenderer.lineHeight);
            }
        }
    }
    
    @Override
    public void setEditable(boolean enabled) {
        super.setEditable(enabled);
        pseudoIsEnabled = enabled;
    }
    
    @Override
    public void setTextColor(int color) {
        super.setTextColor(color);
        pseudoEnabledColor = color;
    }
    
    @Override
    public void setTextColorUneditable(int color) {
        super.setTextColorUneditable(color);
        pseudoDisabledColor = color;
    }
    
    @Override
    public void setFocused(boolean isFocused) {
        if (isFocused && !isFocused()) {
            pseudoCursorCounter = 0;
        }
        super.setFocused(isFocused);
    }
    
    @Override
    public void setBordered(boolean enableBackgroundDrawing) {
        super.setBordered(enableBackgroundDrawing);
        pseudoEnableBackgroundDrawing = enableBackgroundDrawing;
    }
    
    @Override
    public void setMaxLength(int length) {
        super.setMaxLength(length);
        pseudoMaxStringLength = length;
    }
    
    @Override
    public void tick() {
        super.tick();
        pseudoCursorCounter++;
    }
    
    @Override
    public void setHighlightPos(int position) {
        super.setHighlightPos(position);
        int i = getValue().length();
        pseudoSelectionEnd = MathHelper.clamp(position, 0, i);
        if (fontRenderer != null) {
            if (pseudoLineScrollOffset > i) {
                pseudoLineScrollOffset = i;
            }
            
            int j = getInnerWidth();
            String s = fontRenderer.plainSubstrByWidth(getValue().substring(this.pseudoLineScrollOffset), j, false);
            int k = s.length() + pseudoLineScrollOffset;
            if (pseudoSelectionEnd == pseudoLineScrollOffset) {
                pseudoLineScrollOffset -= fontRenderer.plainSubstrByWidth(getValue(), j, true).length();
            }
            
            if (pseudoSelectionEnd > k) {
                pseudoLineScrollOffset += pseudoSelectionEnd - k;
            } else if (pseudoSelectionEnd <= pseudoLineScrollOffset) {
                pseudoLineScrollOffset -= pseudoLineScrollOffset - pseudoSelectionEnd;
            }
            
            pseudoLineScrollOffset = MathHelper.clamp(pseudoLineScrollOffset, 0, i);
        }
    }
    
    public void setLabel(ITextComponent label) {
        this.label = label;
    }
    
    public void setLabelColor(int labelColor) {
        this.labelColor = labelColor;
    }
    
    @SuppressWarnings("deprecation")
    private void drawSelectionBox(int startX, int startY, int endX, int endY) {
        if (startX < endX)
            startX = GeneralUtils.swap(endX, endX = startX);
        if (startY < endY)
            startY = GeneralUtils.swap(endY, endY = startY);
        if (endX > x + width)
            endX = x + width;
        if (startX > x + width)
            startX = x + width;
        Tessellator tessellator = Tessellator.getInstance();
        BufferBuilder buffer = tessellator.getBuilder();
    
        // Annotated as deprecated but no replacement
        RenderSystem.color4f(0.0F, 0.0F, 255.0F, 255.0F);
        
        RenderSystem.disableTexture();
        RenderSystem.enableColorLogicOp();
        RenderSystem.logicOp(GlStateManager.LogicOp.OR_REVERSE);
        RenderUtils.updateBuffer(buffer, startX, startY, endX, endY);
        tessellator.end();
        RenderSystem.disableColorLogicOp();
        RenderSystem.enableTexture();
    }
}

File: src/main/kotlin/io/github/samarium150/minecraft/mod/structures_compass/client/gui/widget/TransparentTextField.kt
/*
 * Copyright (c) 2022 Samarium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
 */
package io.github.samarium150.minecraft.mod.structures_compass.client.gui.widget

import com.mojang.blaze3d.platform.GlStateManager
import com.mojang.blaze3d.systems.RenderSystem
import io.github.samarium150.minecraft.mod.structures_compass.client.util.drawRect
import io.github.samarium150.minecraft.mod.structures_compass.client.util.update
import io.github.samarium150.minecraft.mod.structures_compass.util.Rect
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.font.TextRenderer
import net.minecraft.client.gui.widget.TextFieldWidget
import net.minecraft.client.render.Tessellator
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.Text
import net.minecraft.util.math.MathHelper

@Environment(EnvType.CLIENT)
class TransparentTextField(
    private val textRenderer: TextRenderer?,
    x: Int,
    y: Int,
    width: Int,
    height: Int,
    private var label: Text
) : TextFieldWidget(textRenderer, x, y, width, height, label) {

    private var labelColor = 0x808080
    private var pseudoEditable = true
    private var pseudoEnableBackgroundDrawing = true
    private var pseudoMaxLength = 32
    private var pseudoLineScrollOffset = 0
    private var pseudoEditableColor = 14737632
    private var pseudoDisabledColor = 7368816
    private var pseudoCursorCounter = 0
    private var pseudoSelectionEnd = 0

    override fun render(matrixStack: MatrixStack, mouseX: Int, mouseY: Int, partialTicks: Float) {
        if (isVisible) {
            if (pseudoEnableBackgroundDrawing) {
                val color = (255.0f * 0.55f).toInt()
                drawRect(x, y, x + width, y + height, color / 2 shl 24)
            }
            val showLabel = !isFocused && text.isEmpty()
            val i = if (showLabel) labelColor else if (pseudoEditable) pseudoEditableColor else pseudoDisabledColor
            val j = cursor - pseudoLineScrollOffset
            var k = pseudoSelectionEnd - pseudoLineScrollOffset
            val text = if (showLabel) label.string else text
            val s = textRenderer!!.trimToWidth(text.substring(pseudoLineScrollOffset), getWidth())
            val flag = j >= 0 && j <= s.length
            val flag1 = isFocused && pseudoCursorCounter / 6 % 2 == 0 && flag
            val l = if (pseudoEnableBackgroundDrawing) x + 4 else x
            val i1 = if (pseudoEnableBackgroundDrawing) y + (height - 8) / 2 else y
            var j1 = l
            if (k > s.length) {
                k = s.length
            }
            if (s.isNotEmpty()) {
                val s1 = if (flag) s.substring(0, j) else s
                j1 = textRenderer.drawWithShadow(matrixStack, s1, l.toFloat(), i1.toFloat(), i)
            }
            val flag2 = cursor < getText().length || getText().length >= pseudoMaxLength
            var k1 = j1
            if (!flag)
                k1 = if (j > 0) l + width else l
            else if (flag2) {
                k1 = j1 - 1
                --j1
            }
            if (s.isNotEmpty() && flag && j < s.length)
                textRenderer.drawWithShadow(matrixStack, s.substring(j), j1.toFloat(), i1.toFloat(), i)

            if (flag1)
                if (flag2)
                    drawRect(k1, i1 - 1, k1 + 1, i1 + 1 + textRenderer.fontHeight, -3092272)
                else
                    textRenderer.drawWithShadow(matrixStack, "_", k1.toFloat(), i1.toFloat(), i)

            if (k != j) {
                val l1 = l + textRenderer.getWidth(s.substring(0, k))
                drawSelectionBox(k1, i1 - 1, l1 - 1, i1 + 1 + textRenderer.fontHeight)
            }
        }
    }

    override fun setEditable(editable: Boolean) {
        super.setEditable(editable)
        pseudoEditable = editable
    }

    override fun setEditableColor(color: Int) {
        super.setEditableColor(color)
        pseudoEditableColor = color
    }

    override fun setUneditableColor(color: Int) {
        super.setUneditableColor(color)
        pseudoDisabledColor = color
    }

    public override fun setFocused(isFocused: Boolean) {
        if (isFocused && !isFocused()) {
            pseudoCursorCounter = 0
        }
        super.setFocused(isFocused)
    }

    override fun setDrawsBackground(drawsBackground: Boolean) {
        super.setDrawsBackground(drawsBackground)
        pseudoEnableBackgroundDrawing = drawsBackground
    }

    override fun setMaxLength(length: Int) {
        super.setMaxLength(length)
        pseudoMaxLength = length
    }

    override fun tick() {
        super.tick()
        pseudoCursorCounter++
    }

    override fun setSelectionEnd(position: Int) {
        super.setSelectionEnd(position)
        val i = text.length
        pseudoSelectionEnd = MathHelper.clamp(position, 0, i)
        if (textRenderer != null) {
            if (pseudoLineScrollOffset > i) {
                pseudoLineScrollOffset = i
            }
            val j = innerWidth
            val s = textRenderer.trimToWidth(text.substring(pseudoLineScrollOffset), j, false)
            val k = s.length + pseudoLineScrollOffset
            if (pseudoSelectionEnd == pseudoLineScrollOffset) {
                pseudoLineScrollOffset -= textRenderer.trimToWidth(text, j, true).length
            }
            if (pseudoSelectionEnd > k) {
                pseudoLineScrollOffset += pseudoSelectionEnd - k
            } else if (pseudoSelectionEnd <= pseudoLineScrollOffset) {
                pseudoLineScrollOffset -= pseudoLineScrollOffset - pseudoSelectionEnd
            }
            pseudoLineScrollOffset = MathHelper.clamp(pseudoLineScrollOffset, 0, i)
        }
    }

    @Suppress("unused")
    fun setLabel(label: Text) {
        this.label = label
    }

    @Suppress("unused")
    fun setLabelColor(labelColor: Int) {
        this.labelColor = labelColor
    }

    private fun drawSelectionBox(startX: Int, startY: Int, endX: Int, endY: Int) {
        val (sanitizedStartX, sanitizedStartY, sanitizedEndX, sanitizedEndY) =
            Rect(startX, startY, endX, endY).sanitize()
        val tessellator = Tessellator.getInstance()
        val buffer = tessellator.buffer
        RenderSystem.setShaderColor(0.0f, 0.0f, 255.0f, 255.0f)
        RenderSystem.disableTexture()
        RenderSystem.enableColorLogicOp()
        RenderSystem.logicOp(GlStateManager.LogicOp.OR_REVERSE)
        buffer.update(sanitizedStartX, sanitizedStartY, sanitizedEndX, sanitizedEndY)
        tessellator.draw()
        RenderSystem.disableColorLogicOp()
        RenderSystem.enableTexture()
    }
}
